cc/dual_net/batching_dual_net.cc

```diff
@@ -0,0 +1,176 @@
+#include "cc/dual_net/batching_dual_net.h"
+
+#include <future>
+#include <queue>
+
+#include "absl/memory/memory.h"
+#include "absl/synchronization/mutex.h"
+#include "absl/synchronization/notification.h"
+#include "cc/check.h"
+
+namespace minigo {
+namespace {
+class BatchingService {
+  struct InferenceData {
+    std::vector<const DualNet::BoardFeatures*> features;
+    std::vector<DualNet::Output*> outputs;
+    std::string* model;
+    absl::Notification* notification;
+  };
+
+ public:
+  explicit BatchingService(std::unique_ptr<DualNet> dual_net)
+      : dual_net_(std::move(dual_net)),
+        num_clients_(0),
+        queue_counter_(0),
+        run_counter_(0),
+        num_runs_(0) {}
+
+  ~BatchingService() {
+    std::cerr << "Ran " << num_runs_ << " batches with an average size of "
+              << static_cast<float>(run_counter_) / num_runs_ << ".\n";
+  }
+
+  void IncrementClientCount() {
+    absl::MutexLock lock(&mutex_);
+    ++num_clients_;
+  }
+
+  void DecrementClientCount() {
+    absl::MutexLock lock(&mutex_);
+    --num_clients_;
+    MaybeRunBatches();
+  }
+
+  void RunMany(std::vector<const DualNet::BoardFeatures*> features,
```

Comment:
Nit: bad alignment

```
  void RunMany(std::vector<const DualNet::BoardFeatures*> features,
               std::vector<DualNet::Output*> outputs,
               std::string* model) {
```

Reply:
Oops, I forgot to apply clang-format. Done.